### PR TITLE
Move track command

### PR DESCRIPTION
**Pull request by Telluur**
Adds support for moving a track in queue.

### This pull request...
  - [ ] Fixes a bug
  - [x] Introduces a new feature
  - [ ] Improves an existing feature
  - [ ] Boosts code quality or performance

### Description
Adds `?movetrack <from> <to>` command to move the queued item `<from>` to the specified position `<to>`.

### Purpose
Adds more control to the queue. Implemented as DJ command.

### Notes from the author
Implemented this in my own fork which uses a different code style (braces etc). I manually tried to get it in line with the upstream (your) repository, but an automated refactor might be wise.

### CODE REVIEW

File: src/main/java/com/jagrosh/jmusicbot/JMusicBot.java [hunk 1]
```diff
@@ -118,6 +118,7 @@ public static void main(String[] args)
                         new PlaynextCmd(bot, config.getLoading()),
                         new RepeatCmd(bot),
                         new SkiptoCmd(bot),
+                        new MoveTrackCmd(bot),
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/JMusicBot.java]: Commands should be sorted alphabetically

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Command that provides users the ability to move a track in the playlist.
+ */
+public class MoveTrackCmd extends DJCommand
+{
+
+    private static final Pattern PATTERN = Pattern.compile("^(\\d+)\\s+(\\d+)$");
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: I'm not a huge fan of using a regex here; I'd rather split on whitespace first and then attempt to parse integers (take a look at the other commands that accept numerical inputs)

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    public MoveTrackCmd(Bot bot)
+    {
+        super(bot);
+        this.name = "movetrack";
+        this.help = "Move a track in the current playlist to a different position";
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: minor nitpick but the first word here shouldn't be capitalized

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        int to;
+
+        try
+        {
+            // Validate the args
+            String args = event.getArgs().trim();
+            Matcher matcher = PATTERN.matcher(args);
+            if (!matcher.matches())
+            {
+                event.replyError("That ain't right. Usage: movetrack <from> <to>");
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: language/tone should be consistent with the rest of the bot; no sarcastic replies please!

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                return;
+            }
+
+            from = Integer.parseInt(matcher.group(1));
+            to = Integer.parseInt(matcher.group(2));
+        }
+        catch (NumberFormatException e)
+        {
+            // Should already be caught by the regex but ok..
+            event.replyError("That ain't a number: " + e.getMessage());
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: same thing as above (language/tone), and also the regex wouldn't catch extremely large numbers that can't be parsed by integer

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            return;
+        }
+
+        if (from == to)
+        {
+            event.replySuccess("Wow! That was easy. Great job using this command. You're doing a great job.");
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: language/tone, and also shouldn't be `replySuccess` if this isn't a successful use of the command

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        // Move the track
+        QueuedTrack track = queue.moveItem(from - 1, to - 1);
+        String trackTitle = track.getTrack().getInfo().title;
+        String reply = String.format("Moved track '%s' from position %d to %d.", trackTitle, from, to);
+        event.replySuccess(reply);
+    }
+
+    private boolean isAvailablePosition(CommandEvent event, FairQueue<QueuedTrack> queue, int position)
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: as written, this method should probably be static, but I'm not a huge fan of delegating the `event.reply` into a different method; maybe just return true or false and have the reply in the execute method for easier readability of what is happening

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        FairQueue<QueuedTrack> queue = handler.getQueue();
+        if (!isAvailablePosition(event, queue, from) || !isAvailablePosition(event, queue, to))
+        {
+            return;
+        }
+
+        // Move the track
+        QueuedTrack track = queue.moveItem(from - 1, to - 1);
+        String trackTitle = track.getTrack().getInfo().title;
+        String reply = String.format("Moved track '%s' from position %d to %d.", trackTitle, from, to);
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: This should use the same markdown formatting as other commands that provide song titles and positions

Comment (Telluur) [reply src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: Just to clarify, That would mean bold track names, and backticks around the user provided positions?

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java (hunk 7 above)
Comment (Telluur) [reply src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: Agreed, made static and made 2 separate checks in command execution method to provide the user with detailed feedback.

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java (hunk 6 above)
Comment (Telluur) [reply src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: Fixed

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public void doCommand(CommandEvent event)
+    {
+        int from;
+        int to;
+
+        String[] parts = event.getArgs().split("\\s+", 2);
+        if(parts.length < 2)
+        {
+            event.replyError("Please include two valid indexes.'");
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: extra single-quote

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            String reply = String.format("`%d` is not a valid position in the queue!", to);
+            event.replyError(reply);
+            return;
+        }
+
+        // Move the track
+        QueuedTrack track = queue.moveItem(from - 1, to - 1);
+        String trackTitle = track.getTrack().getInfo().title;
+        String reply = String.format("Moved **%s** from position `%d` to `%d.`", trackTitle, from, to);
```
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/MoveTrackCmd.java]: move the period to outside of the backticks

**Comment by jagrosh** [comment]
Thanks for your contribution!

**Comment by Telluur** [comment]
Might want to rebase+squash my three commits into one, if you want to tidy up the commit history. And thank you for this bot. :)

**Comment by jagrosh** [comment]
I don't mind a few extra commits for fixes, and rebase+squash would mess up github's contributor graph

**Pull request merged**